codex/stores/repostore.nim

```diff
@@ -41,126 +45,277 @@ declareGauge(codex_repostore_bytes_reserved, "codex repostore bytes reserved")
 
 const
   DefaultBlockTtl* = 24.hours
-  DefaultQuotaBytes* = 1'u shl 33'u # ~8GB
+  DefaultQuotaBytes* = 8.GiBs
 
 type
-  QuotaUsedError* = object of CodexError
   QuotaNotEnoughError* = object of CodexError
 
   RepoStore* = ref object of BlockStore
     postFixLen*: int
     repoDs*: Datastore
     metaDs*: Datastore
     clock: Clock
-    totalBlocks*: uint            # number of blocks in the store
-    quotaMaxBytes*: uint          # maximum available bytes
-    quotaUsedBytes*: uint         # bytes used by the repo
-    quotaReservedBytes*: uint     # bytes reserved by the repo
+    quotaMaxBytes*: NBytes
+    quotaUsage*: QuotaUsage
+    totalBlocks*: Natural
     blockTtl*: Duration
     started*: bool
 
+  QuotaUsage* = object
+    used: NBytes
+    reserved: NBytes
+
+  BlockMetadata* = object
+    expiry*: SecondsSince1970
+    size*: NBytes
+    refCount*: Natural
+
+  LeafMetadata* = object
+    blkCid: Cid
+    proof: CodexProof
+
   BlockExpiration* = object
     cid*: Cid
-    expiration*: SecondsSince1970
+    expiry*: SecondsSince1970
 
-proc updateMetrics(self: RepoStore) =
-  codex_repostore_blocks.set(self.totalBlocks.int64)
-  codex_repostore_bytes_used.set(self.quotaUsedBytes.int64)
-  codex_repostore_bytes_reserved.set(self.quotaReservedBytes.int64)
+func quotaUsedBytes*(self: RepoStore): NBytes =
+  self.quotaUsage.used
 
-func totalUsed*(self: RepoStore): uint =
+func quotaReservedBytes*(self: RepoStore): NBytes =
+  self.quotaUsage.reserved
+
+func totalUsed*(self: RepoStore): NBytes =
   (self.quotaUsedBytes + self.quotaReservedBytes)
 
-func available*(self: RepoStore): uint =
+func available*(self: RepoStore): NBytes =
   return self.quotaMaxBytes - self.totalUsed
 
-func available*(self: RepoStore, bytes: uint): bool =
+func available*(self: RepoStore, bytes: NBytes): bool =
   return bytes < self.available()
 
-proc encode(cidAndProof: (Cid, CodexProof)): seq[byte] =
-  ## Encodes a tuple of cid and merkle proof in a following format:
-  ## | 8-bytes | n-bytes | remaining bytes |
-  ## |    n    |   cid   |      proof      |
-  ##
-  ## where n is a size of cid
-  ##
-  let
-    (cid, proof) = cidAndProof
-    cidBytes = cid.data.buffer
-    proofBytes = proof.encode
-    n = cidBytes.len
-    nBytes = n.uint64.toBytesBE
-
-  @nBytes & cidBytes & proofBytes
-
-proc decode(_: type (Cid, CodexProof), data: seq[byte]): ?!(Cid, CodexProof) =
-  let
-    n = uint64.fromBytesBE(data[0..<sizeof(uint64)]).int
-    cid = ? Cid.init(data[sizeof(uint64)..<sizeof(uint64) + n]).mapFailure
-    proof = ? CodexProof.decode(data[sizeof(uint64) + n..^1])
-  success((cid, proof))
-
-proc decodeCid(_: type (Cid, CodexProof), data: seq[byte]): ?!Cid =
-  let
-    n = uint64.fromBytesBE(data[0..<sizeof(uint64)]).int
-    cid = ? Cid.init(data[sizeof(uint64)..<sizeof(uint64) + n]).mapFailure
-  success(cid)
+proc encode(t: Cid): seq[byte] = t.data.buffer
```

Comment:
this file is already quite busy, can we perhaps split this helpres out to their own unit?

Reply:
Except for `Cid` all `encode/decode` procs are defined here because all corresponding types are also defined here.

Reply:
Then it's probably good to extract the types to their own `types` module and share across both, this is the commonly accepted approach.

Reply:
This file needs to be split into several functional units that group related functionality. 

I suggest the following structure: 

- `respostore/types.nim` - common types shared by all other functional unites
- `repostore/coders.nim` - encoding/decoding primitives
- `repostore/operations.nim` - the update/get/modify operation
- `repostore/repostore.nim/store.nim` - the actual repostore functionality
- `repostore.nim` - top level module that exports the required public functinal units/modules

This structure would make the changes easier to track and compare with previous versions. It's quite hard to follow whats going on right now and the file is already quite large and busy with unrelated functionality.

Reply:
Ok